### PR TITLE
address Google OAuth2 issues where user didn't grant us the `https://www.googleapis.com/auth/calendar.events.readonly` scope

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Follow up PR to https://github.com/grafana/oncall/pull/4792

Basically if when communicating with Google Calendar's API we encounter an HTTP 403, or the Google client throws a `google.auth.exceptions.RefreshError` this means one of three things:
1. the refresh token we have persisted for the user is missing the `https://www.googleapis.com/auth/calendar.events.readonly` scope (HTTP 403)
2. the Google user has been deleted (`google.auth.exceptions.RefreshError`)
3. the refresh token has expired (`google.auth.exceptions.RefreshError`)

To prevent scenario 1 above from happening in the future we now will check that the token has been granted the required scopes. If the user doesn't grant us all the necessary scopes, we will show them an error message in the UI:
https://www.loom.com/share/0055ef03192b4154b894c2221cecbd5f

For tokens that were granted prior to this PR and which are missing the required scope, we will show the user a dismissible warning banner in the UI letting them know that they will need to reconnect their account and grant us the missing permissions (see [this second demo video](https://www.loom.com/share/bf2ee8b840864a64893165370a892bcd) showing this).

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] Added the relevant release notes label (see labels prefixed w/ `release:`). These labels dictate how your PR will
    show up in the autogenerated release notes.

### CODE REVIEW

**Review by matiasb — APPROVED** [review]

File: grafana-plugin/src/containers/Alerts/Alerts.tsx [hunk 1]
```diff
@@ -54,17 +57,33 @@ export const Alerts = function () {
...
+    /**
+     * TODO: if we don't dereference currentUserPk then for some reason the @computed portion of
+     * UserStore.currentUser doesn't recalculate and will just be stuck on returning undefined..
+     *
+     * If we dereference currentUserPk here, even if we don't use it.. things just seem to work
+     * (what is this mobx wizardry?)
+     *
+     * Seems to be related to this https://stackoverflow.com/questions/77724466/mobx-computed-not-updating
+     */
+    userStore: { currentUser, currentUserPk },
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/Alerts/Alerts.tsx:69]: @grafana/grafana-oncall-frontend for the life of me I can't figure this out 😆

it works, but if I delete the dereferenced `currentUserPk` (which isn't actually used in this scope), it stops working

File: grafana-plugin/src/containers/UserSettings/UserSettings.module.css [hunk 2]
```diff
@@ -8,3 +8,18 @@
...
+  gap: 10px;
+
+  &:empty {
+    display: none;
+  }
+}
+
+.alert {
+  margin: 0;
+}
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/UserSettings/UserSettings.module.css:25]: @grafana/grafana-oncall-frontend this was essentially copy-pasta'd from [`grafana-plugin/src/containers/Alerts/Alerts.module.scss`](https://github.com/grafana/oncall/blob/dev/grafana-plugin/src/containers/Alerts/Alerts.module.scss#L1-L19), not sure if we want to extract this out somehow?

File: grafana-plugin/src/containers/Alerts/Alerts.tsx (hunk 1 above)
Comment (joeyorlando) [reply grafana-plugin/src/containers/Alerts/Alerts.tsx:69]: this actually breaks the frontend build for now

https://github.com/grafana/oncall/actions/runs/10393286285/job/28780417264?pr=4802#step:4:15

File: engine/apps/google/tasks.py [hunk 3]
```diff
@@ -118,5 +128,11 @@ def sync_out_of_office_calendar_events_for_user(google_oauth2_user_pk: int) -> N
 
 @shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True)
 def sync_out_of_office_calendar_events_for_all_users() -> None:
-    for google_oauth2_user in GoogleOAuth2User.objects.filter(user__organization__deleted_at__isnull=True):
+    # some existing tokens may not have all the required scopes, lets skip these
+    tokens_containing_required_scopes = GoogleOAuth2User.objects.filter(
+        *[Q(oauth_scope__contains=scope) for scope in constants.REQUIRED_OAUTH_SCOPES],
+        user__organization__deleted_at__isnull=True,
+    )
+
```
Comment (mderynck) [inline engine/apps/google/tasks.py:136]: Would it be useful to have a `x/total` log message here so we know how we are doing for users correcting not having the required scopes?

Comment (joeyorlando) [reply engine/apps/google/tasks.py:136]: Good idea 👍 added in https://github.com/grafana/oncall/pull/4802/commits/6ec1cb13589616bdcb42ee85db864f7bb6f81186

File: grafana-plugin/src/containers/Alerts/Alerts.tsx [hunk 4]
```diff
@@ -164,4 +190,24 @@ export const Alerts = function () {
...
+   */
+  function showCurrentUserGoogleOAuth2TokenIsMissingScopes(): boolean {
+    return Boolean(
+      currentUserPk &&
+        currentUser &&
+        currentUser.has_google_oauth2_connected &&
+        currentUser.google_oauth2_token_is_missing_scopes &&
+        !getItem(AlertID.USER_GOOGLE_OAUTH2_TOKEN_MISSING_SCOPES)
+    );
+  }
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/Alerts/Alerts.tsx:212]: @grafana/grafana-oncall-frontend for the life of me I can't figure this out 😆

it works, but if I delete the reference to `currentUserPk` (which isn't actually needed), it stops working

**Pull request merged**